src/main/java/com/android/volley/AsyncNetwork.java

```diff
@@ -0,0 +1,124 @@
+
+import androidx.annotation.RestrictTo;
+import com.android.volley.toolbox.AsyncHttpStack;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+/** An asynchronous implementation of {@link Network} to perform requests. */
+public abstract class AsyncNetwork implements Network {
+    private final AsyncHttpStack mAsyncStack;
+    private ExecutorService mBlockingExecutor;
+    private ExecutorService mNonBlockingExecutor;
+
+    protected AsyncNetwork(AsyncHttpStack stack) {
+        mAsyncStack = stack;
+    }
+
+    public interface OnRequestComplete {
+        void onSuccess(NetworkResponse networkResponse);
+
+        void onError(VolleyError volleyError);
+    }
+
+    /**
+     * Non-blocking method to perform the specified request.
+     *
+     * @param request Request to process
+     * @param callback to be called once NetworkResponse is received
+     */
+    public abstract void performRequest(Request<?> request, OnRequestComplete callback);
+
+    /**
+     * Blocking method to perform network request.
+     *
+     * @param request Request to process
+     * @return response retrieved from the network
+     * @throws VolleyError in the event of an error
+     */
+    @Override
+    public NetworkResponse performRequest(Request<?> request) throws VolleyError {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicReference<NetworkResponse> response = new AtomicReference<>();
+        final AtomicReference<VolleyError> error = new AtomicReference<>();
+        performRequest(
+                request,
+                new OnRequestComplete() {
+                    @Override
+                    public void onSuccess(NetworkResponse networkResponse) {
+                        response.set(networkResponse);
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public void onError(VolleyError volleyError) {
+                        error.set(volleyError);
+                        latch.countDown();
+                    }
+                });
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            VolleyLog.e(e, "while waiting for CountDownLatch");
+            Thread.currentThread().interrupt();
+            throw new VolleyError(e);
+        }
+
+        if (response.get() != null) {
+            return response.get();
+        } else if (error.get() != null) {
+            throw error.get();
+        } else {
+            throw new VolleyError("Neither response entry was set");
+        }
+    }
+
+    /**
+     * This method sets the non blocking executor to be used by the stack for non-blocking tasks.
+     * This method must be called before performing any requests.
+     */
+    @RestrictTo({RestrictTo.Scope.LIBRARY_GROUP})
+    public void setNonBlockingExecutorForStack(ExecutorService executor) {
```

Comment:
"setNonBlockingExecutor"? Not sure what the "ForStack" is for here since it's no different from what setBlockingExecutor is doing. This executor could be used by the network too.

Reply:
My reasoning was that getNonBlockingExecutor() isn't called at all by BasicAsyncNetwork. But you're right that it could be used by the network.